commons/zenoh-shm/src/api/protocol_implementations/posix/posix_shared_memory_provider_backend.rs

```diff
@@ -0,0 +1,284 @@
+    }
+
+    /// Construct layout in-place using arguments
+    #[zenoh_macros::unstable_doc]
+    pub fn with_layout_args(
+        self,
+        size: usize,
+        alignment: AllocAlignment,
+    ) -> ZResult<LayoutedPosixSharedMemoryProviderBackendBuilder<MemoryLayout>> {
+        let layout = MemoryLayout::new(size, alignment)?;
+        Ok(LayoutedPosixSharedMemoryProviderBackendBuilder { layout })
+    }
+
+    /// Construct layout in-place from size (default alignment will be used)
+    #[zenoh_macros::unstable_doc]
+    pub fn with_size(
+        self,
+        size: usize,
+    ) -> ZResult<LayoutedPosixSharedMemoryProviderBackendBuilder<MemoryLayout>> {
+        let layout = MemoryLayout::new(size, AllocAlignment::default())?;
+        Ok(LayoutedPosixSharedMemoryProviderBackendBuilder { layout })
+    }
+}
+
+pub struct LayoutedPosixSharedMemoryProviderBackendBuilder<Layout: Borrow<MemoryLayout>> {
+    layout: Layout,
+}
+
+impl<Layout: Borrow<MemoryLayout>> LayoutedPosixSharedMemoryProviderBackendBuilder<Layout> {
+    /// try to create PosixSharedMemoryProviderBackend
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> ZResult<PosixSharedMemoryProviderBackend> {
+        PosixSharedMemoryProviderBackend::new(self.layout.borrow())
+    }
+}
+
+/// A backend for SharedMemoryProvider based on POSIX shared memory.
+/// This is the default general-purpose backed shipped with Zenoh.
+#[zenoh_macros::unstable_doc]
+pub struct PosixSharedMemoryProviderBackend {
+    available: AtomicUsize,
+    segment: PosixSharedMemorySegment,
+    free_list: Mutex<BinaryHeap<Chunk>>,
+    alignment: AllocAlignment,
+}
+
+impl PosixSharedMemoryProviderBackend {
+    /// Get the builder to construct a new instance
+    #[zenoh_macros::unstable_doc]
+    pub fn builder() -> PosixSharedMemoryProviderBackendBuilder {
+        PosixSharedMemoryProviderBackendBuilder
+    }
+
+    fn new(layout: &MemoryLayout) -> ZResult<Self> {
+        let segment = PosixSharedMemorySegment::create(layout.size())?;
+
+        let mut free_list = BinaryHeap::new();
+        let root_chunk = Chunk {
+            offset: 0,
+            size: layout.size(),
+        };
+        free_list.push(root_chunk);
+
+        tracing::trace!(
+            "Created PosixSharedMemoryProviderBackend id {}, layout {:?}",
+            segment.segment.id(),
+            layout
+        );
+
+        Ok(Self {
+            available: AtomicUsize::new(layout.size()),
+            segment,
+            free_list: Mutex::new(free_list),
+            alignment: layout.alignment(),
+        })
+    }
+}
+
+impl SharedMemoryProviderBackend for PosixSharedMemoryProviderBackend {
+    fn alloc(&self, layout: &MemoryLayout) -> ChunkAllocResult {
+        tracing::trace!("PosixSharedMemoryProviderBackend::alloc({:?})", layout);
+
+        let required_len = layout.size();
+
+        if self.available.load(Ordering::Relaxed) >= required_len {
```

Comment:
In general I believe it's more readable having an early return instead of a big block of nested code.
E.g.:
```rust 
if self.available.load(Ordering::Relaxed) < required_len {
  tracing::trace!( "PosixSharedMemoryProviderBackend does not have sufficient free memory to allocate {:?}, try de-fragmenting!", layout);
  return Err(ZAllocError::OutOfMemory);
}

... code to handle Ok case ...
```